### ISSUE TITLE
memory inflation after cuDF query

### ISSUE THREAD

**Issue by miroenev** [rapidsai/cudf#392]
Running a  [cudf.query()](https://github.com/rapidsai/cudf/blob/a5ff0806f5873aa13a9b7fe95db5ceb26cf07988/cudf/dataframe.py#L1201) results in increased memory usage proportional to the size of the dataframe.

**Steps to reproduce:**

1. Generate random dataset in numpy and import into cudf via pandas
Random Dataset = 100M rows, 5 columns [ float32 ]
```
nRows = int(1e8)
nCols = 5
dataNumpy = np.asfortranarray( np.random.rand( nRows, nCols) ) # https://github.com/rapidsai/cudf/issues/209
colNames = ['col'+str(iCol) for iCol in range(nCols)]

pandasDF = pd.DataFrame( data = dataNumpy, columns = colNames, dtype = np.float32)
cudaDF = cudf.dataframe.DataFrame.from_pandas( pandasDF )
```
2. check memory consumption before cudf.query
3.329490944 GB used [on GPU# 0] before filtering :: dataframe rows = 100000000
3. run query [ which does not change dataframe size ]
```cudaDF = cudaDF.query('col0 > 0')```
4. check memory consumption after query
6.5339392 GB used [on GPU# 0] after filtering :: dataframe rows = 100000000

**Additional notes:**
Behavior  is consistent when [RAPIDS Memory Manager](https://github.com/rapidsai/cudf/blob/master/libgdf/doc/RAPIDS_Memory_Manager.md) is enabled or disabled.
Most of the time this issue is only observed on the first query/filtering call of each dataframe.

**Version & Environment details:**
OS: Ubuntu LTS 16.04 -- Kernel: 4.4.0-138-generic
GPU: Tesla V100 16GB -- Driver: 410.73
Arrow: 0.10.0
CMake: 3.5.1
CUDA: 9.2.148
gcc/g++: 5.4.0
Numpy: 1.14.3
Pandas: 0.20.3
Python: 3.6.6
cuML: 0.20.0

[Dockerfile.txt](https://github.com/rapidsai/cudf/files/2593886/Dockerfile.txt) [ attached modified from rapidsai/cudf/Dockerfile ]

**Reproducible Sample:**
[cudf_filter_memory_issue_repro.ipynb.txt](https://github.com/rapidsai/cudf/files/2593841/cudf_filter_memory_issue_repro.ipynb.txt) notebook is attached (remove .txt extension as .ipynb extension was not allowed by github ).

![issue_notebook_snapshot](https://user-images.githubusercontent.com/8701332/48687075-36079380-eb74-11e8-905d-b7e56c11eb3b.PNG)

**Comment by amotupal** [comment]
This is a huge blocker for me. I have 30M records and the before query it has 2335MiB where as after query it is using 9153MiB.

**Comment by harrism** [comment]
I did some digging and this doesn't appear to be an RMM problem. I suspect it's somewhere in the allocations that cuDF does, or in the way that Numba does GC... @kkraus14 can you assign someone to dig on the Python side?

**Comment by amotupal** [comment]
If you want something to reproduce, I can send my jupyter notebook in the PM. Thanks.

**Comment by pearu** [comment]
From tracking the CPU virtual memory usage of a python process, I have found the following:
1. virtual memory change does not depend on the size of data
2. calling `from_pandas` causes a 13 GB increase in the virtual memory usage (the leak)
3. calling `query` does not increase virtual memory usage
Note that only CPU memory is tracked here but the result 2. seems to be weird.

**Comment by pearu** [comment]
The 13 GB increase comes from the first call to `rmmAlloc` (here the Python code turns to C++).
While the first call allocates device memory using cudaMalloc, it is unclear why the host virtual memory usage increases.

When switching to using memory pool allocator via
```
from librmm_cffi import librmm_config as rmm_cfg
rmm_cfg.use_pool_allocator = True
rmm_cfg.initial_pool_size = 0
```
then the virtual memory increase becomes 30MB (which is much smaller than 13 GB).

Increasing `initial_pool_size` from 0 to 4.8 GB, the increase stays constant (30MB) but
at 5GB there is a jump to 4GB host virtual memory usage.
My GPU device (P2000) has 5GB, using CUDA 9.2.

Can anyone explain the host memory increase as described above?

**Comment by pearu** [comment]
I now tracked also GPU memory usage when changing the rmm parameters with the following conclusions:
1. With `use_pool_allocator = True` and `initial_pool_size < total device memory` there are no leaks neither in host nor in device memory
2. With `use_pool_allocator = True` and `initial_pool_size > total device memory` there are apparent leaks: host 176 MB and device 122 MB
3. With  `use_pool_allocator = False` there are apparent leaks: host 80 MB and device 14 MB

The test array size in all cases is about 38MB.

**Comment by harrism** [comment]
I'm not clear why you are studying CPU virtual memory usage, or testing with the pool allocator. The issue as reported demonstrates *GPU* (not CPU virtual) memory inflation when pool is disabled.

BTW,  `initial_pool_size = 0` (the default when `use_pool_allocator = True`) creates a pool with size equal to half the total GPU memory size.

**Comment by kkraus14** [comment]
I'm exploring this, after calling `.query()` the first time and then forcing numba to run garbage collection: `cuda.current_context().deallocations.clear()` it still doesn't free the memory, so we're likely  erroneously holding a reference somewhere.

**Comment by pearu** [comment]
@harrism yes, the issue is for GPU, but it seemed that the abnormalities with host memory are related.

**Comment by kkraus14** [comment]
The issue is unrelated to querying, minimal reproducer:
```python
import numpy as np
import pandas as pd
import cudf

nRows = int(1e8)
nCols = 2
dataNumpy = np.asfortranarray( np.random.rand( nRows, nCols) ) # https://github.com/rapidsai/cudf/issues/209
colNames = ['col'+str(iCol) for iCol in range(nCols)]

pandasDF = pd.DataFrame( data = dataNumpy, columns = colNames, dtype = np.float32)
cudaDF = cudf.dataframe.DataFrame.from_pandas( pandasDF )

boolmask = cudf.Series(np.random.randint(1, 2, len(cudaDF)).astype('bool'))

cudaDF = cudaDF[boolmask]
```

**Comment by kkraus14** [comment]
I'm 95% sure what's happening is that in the original DataFrame creation each column shares a reference to the same index, but after calling the mask there's a new index created for each column and the memory blows up because of it.

**Comment by kkraus14** [comment]
Confirmed, before boolean masking:
```
In [3]: print(cudaDF.index._values.data.mem)                                                                                               
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f8abf0080>

In [4]: print(cudaDF['col0'].index._values.data.mem)                                                                                       
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f8abf0080>

In [5]: print(cudaDF['col1'].index._values.data.mem)                                                                                       
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f8abf0080>
```
After boolean masking:
```
In [7]: print(cudaDF.index._values.data.mem)                                                                                               
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f48087f28>

In [8]: print(cudaDF['col0'].index._values.data.mem)                                                                                       
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f48087f28>

In [9]: print(cudaDF['col1'].index._values.data.mem)                                                                                       
<numba.cuda.cudadrv.devicearray.DeviceNDArray object at 0x7f7f493b5710>
```

**Comment by thomcom** [comment]
In your above csv there are two leaks @kkraus14, `0x7f0364000000` and `0x7f02a4000000`.

**Comment by thomcom** [comment]
There's two leaks in your .csv above, each 800,000,000b.
```
[Address         0x7f0364000000
Size (bytes)         800000000
Name: 105, dtype: object, Address         0x7f02a4000000
Size (bytes)         800000000
Name: 153, dtype: object]
```

**Comment by dantegd** [comment]
@thomcom isn't that is exactly what @kkraus14 is describing: `a new index created for each column `? or am I misreading your comment?

**Comment by thomcom** [comment]
I think it supports what he's saying. Copies always default to deep - it strikes me we just need to identify the spot where the index is deep copied in this instance and make it a shallow copy instead.

**Comment by thomcom** [comment]
@kkraus14 I think the loc that is losing the Index here is one of mine, so I'm digging pretty hard on this.

**Comment by kkraus14** [comment]
Resolved by #782.
